### ISSUE TITLE
VSCode with "Microsoft Python Analysis Engine" makes system unresponsive

### ISSUE THREAD

**Issue by joseph-long** [microsoft/python-language-server#1460]
Potentially related to https://github.com/microsoft/python-language-server/issues/1298 or https://github.com/microsoft/python-language-server/issues/832 or https://github.com/microsoft/vscode-python/issues/4990.

I'm trying to use VSCode on a resource constrained machine (ASUS Chromebook C302 with 4 GB RAM, running Kubuntu 19.04 with swap and compressed RAM to try and alleviate memory pressure). VSCode prompted me to increase the number of inotify watches, as it always does on a new install of Linux, so I did. Shortly after I noticed the status bar showing messages about 1000s of files to analyze. Shortly after _that_ my entire system became unresponsive, as it tends to do when memory and swap are exhausted.

 I can only assume it's analyzing my entire conda environment's site packages, since the file I was editing and indeed the project I was working on only use the standard library (except for one "import IPython" in a demo program).

I switched to using the `jedi` backend for the Python extension, and have been able to edit my project again.

Happy to perform tests or provide more info if someone tells me what to look for!

**Comment by MikhailArkhipov** [comment]
It is not analyzing the entire environment, but `import tensorflow` equals to 3000+ files library. This may be a problem for smaller machines. https://github.com/microsoft/python-language-server/issues/472 will eventually help, but not for all possible cases.

**Comment by talolard** [comment]
![image](https://user-images.githubusercontent.com/5352830/66997337-8bc12480-f0d2-11e9-9aef-4d3224d27f26.png)
Am getting similar problems when using spacy and requests in the new jupyter notebook env

**Comment by jakebailey** [comment]
The original issue was more about resource usage in general; if what you're observing doesn't eventually stop (with the analysis finishing), then that'd be a bug which should be reported separately. Note that the default behavior of `htop` is to show the threads, so the number of things showing there doesn't surprise me (note the same CPU percentage for all, since it's the same process), but the memory usage is certainly not good.

**Comment by braineo** [comment]
![image](https://user-images.githubusercontent.com/3591338/70426476-71923a00-1ab6-11ea-8c85-477479c9292d.png)

Hmm I wonder where to start report with though. Tried compiling 3ad819639ce551c5e3cabc734b507696f3cfd393 but no luck. I need to kill it from time to time otherwise it just halts my machine :(

**Comment by karantza** [comment]
Commenting because I also see an issue with my system becoming unresponsive, in my case I've confirmed it's due to a huge number of threads being created. I'm on Ubuntu, and my system is actually running out of process handles. I'm currently looking at >10,000 python language server threads. If I don't kill the process, eventually even my shell fails to work.

`➜ ps -efL | grep Python.LanguageServer | wc -l
10565`

**Comment by jakebailey** [comment]
That's very strange, I'm not sure exactly what could be making all those threads. The analysis has a concurrent task limit, so maybe it's workspace/library indexing. I'd suggest using lldb to try and grab some stack traces, but if there really are 10k threads, that might get unruly.

**Comment by agalazis** [comment]
> It is not analyzing the entire environment, but `import tensorflow` equals to 3000+ files library. This may be a problem for smaller machines. #472 will eventually help, but not for all possible cases.

bardon my ignorance but why is that? why doesn't it analyse just tesorflow.x (and its dependencies) if and only if/when   tensroflow.x is used?

**Comment by jakebailey** [comment]
Doing so is very complicated, as everything that happens previously can affect things after. Especially for libraries like tensorflow, whose toplevel module (i.e. what you get when you type `import tensorflow`) is just a bunch of `from ... import *`, where we have to resolve everything and reconstruct the global scope. Tensorflow 1.0's `__init__.py`: https://github.com/tensorflow/tensorflow/blob/1.8.0/tensorflow/python/__init__.py

Along with a bunch of other rules about submodules (which when imported in one module mean they get added for _all_ modules), forward references, and then compounded massively by optional imports (try/except import blocks) that massively change how difficult a library is to analyze depending on what a user happens to have installed (i.e. giant conda environments with many unused packages that we need to have participate in the analysis).

**Comment by npexciton** [comment]
Agreed. It's so complicated a language that static analysis is more than impossible. I even tend to believe Python will sooner or later be replaced by something else that has predictable behavior.

I'm just curious about one thing: what editors do people use to write Python programs? Does there exist an editor in the world that just works? vim? PyCharm?

**Comment by jakebailey** [comment]
I said complicated, not impossible. Like any language, there are behaviors that take time to get right when working statically. There has been work in this area to get lazy stuff working.

This issue is about the resource consumption of the language server, what appears to be a bug with too many threads being created. I would prefer to not hijack this issue with unrelated items.
